### ISSUE TITLE
Error when model created in unit test

### ISSUE THREAD

**Issue by leewillis77** [spatie/laravel-activitylog#134]
I have a class set up to log activity, and the description includes the causer's name e.g.

```php
class Like extends Model
{
    use LogsActivity;

    public function getDescriptionForEvent(string $eventName): string
    {
        return ':causer.name liked the walk "' . $this->walk->name . '"';
    }
}
```

When I create an instance of the Like model in a unit test it fails because there is no logged in user to put as the causer, and the evaluation of :causer.name fails:

```
Error: Call to a member function toArray() on null

/vendor/spatie/laravel-activitylog/src/ActivityLogger.php:188
/vendor/spatie/laravel-activitylog/src/ActivityLogger.php:191
/vendor/spatie/laravel-activitylog/src/ActivityLogger.php:143
```

Should the package cope with this? Should I be doing something differently in my tests?

Any advice gratefully received!

**Comment by freekmurze** [comment]
Could you give us some more code to reproduce the issue? How are you building up and saving your `Like` model? How did you set up your model?

**Comment by leewillis77** [comment]
Sure  - no problem. I've put together a simple fresh Laravel project with stripped down Model and test. Here's how you can reproduce the issue:

```
$ git clone git@github.com:leewillis77/activitylog-issue.git
$ cd activitylog-issue
$ composer install
$ php artisan vendor:publish --provider="Spatie\Activitylog\ActivitylogServiceProvider" --tag="migrations"
$ cp .env.example .env
```
Configure some database params in .env - the DB will be cleared, so make sure its a dummy one, then run the migrations
```
$ php artisan migrate
```
Then run phpunit to run the sample test (tests/Feature/Issue134Test.php) to show the issue. 
```
$ phpunit
```
You should get output like the following:
```
There was 1 error:

1) Tests\Feature\Issue134Test::testExample
Error: Call to a member function toArray() on null

/private/tmp/activitylog-issue/vendor/spatie/laravel-activitylog/src/ActivityLogger.php:187
/private/tmp/activitylog-issue/vendor/spatie/laravel-activitylog/src/ActivityLogger.php:190
/private/tmp/activitylog-issue/vendor/spatie/laravel-activitylog/src/ActivityLogger.php:143
/private/tmp/activitylog-issue/vendor/spatie/laravel-activitylog/src/Traits/LogsActivity.php:37
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php:334
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php:199
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php:172
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasEvents.php:142
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:660
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:508
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:824
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1316
/private/tmp/activitylog-issue/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1328
/private/tmp/activitylog-issue/tests/Feature/Issue134Test.php:25
```

You also get the same issue if you try and create an instance of the model within `artisan tinker`, e.g. 

```
$ php artisan tinker
Psy Shell v0.8.1 (PHP 7.1.1 — cli) by Justin Hileman
>>> App\Like::create(['walk_id' => 1, 'user_id' => 1]);
PHP error:  Call to a member function toArray() on null in /private/tmp/activitylog-issue/vendor/spatie/laravel-activitylog/src/ActivityLogger.php on line 187
```

**Comment by AlexVanderbist** [comment]
Thanks for your comprehensive example project. The package is however intended to work like that.

You could seed and login a test user when running tests:
```php
auth()->login(User::first());
```

Or you could avoid using the activity causer and use the user_id that actually liked the walk:
```php
public function getDescriptionForEvent(string $eventName): string
{
    $user = User::find($this->user_id);
    return "{$user->name} liked the walk.";
}
```
This might be preferred since it doesn't depend on a user being logged in but on the actual user that performed this activity.

**Comment by leewillis77** [comment]
Hi Alex; 

Thanks for the reply. I understand that in the example I quoted the workaround would probably be OK.

However, if you're saying that the library is only intended to log activity when performed by a logged in user I think that should be more explicit. 

Currently, logging activity without being logged in works fine *as long as* you don't try and user the causer replacement tags in the description and there's nothing to suggest that you can only record activity logs for logged in users. On a technical level, the table definition explicitly allows causer_id to be NULL. 

Can I suggest one or other of the following:

1. The library continues to allow causer ID to be null, default values are provided for causer replacement tags, or at least an exception is thrown rather than a PHP Error.

2. If the library does not support NULL causers, then the DB structure should be amended to make causer ID NOT NULL, and activity log records shouldn't be created if the causer would be NULL.

I'd much prefer [1] as I think forcing logs to have a logged in user severely limits the flexibility of the library. However - happy to pull together a PR for either approach if you let me know which you'd prefer.

**Comment by AlexVanderbist** [comment]
That's a good point. I've discussed this with @freekmurze and throwing an exception when the causer can't be replaced is probably the best solution here.

Feel free to create a PR for this, if not I'll create one myself next week. Thanks for your input!

**Comment by leewillis77** [comment]
Thanks Alex. I've created 2 PRs with alternative approaches. Feel free to reject either - I'll note my personal preference for #135 though :)

Thanks

**Comment by freekmurze** [comment]
The newly released `v1.12.1`  included the changes in #135
